Improve quickstart: Add compile section (#2607)
PennyLaneAI/pennylane: doc/introduction/compiling_circuits.rst

into one or more differing circuits (or sequences of operators). For example, such a transformation could
replace a gate type with another, fuse gates, exploit mathematical relations that simplify an observable,
or replace a large circuit by a number of smaller circuits.

Compilation functionality is mostly designed as **transforms**, which you can read up on in the
section on :doc:`inspecting circuits </introduction/inspecting_circuits>`.

Compilation transforms for circuit optimization
-----------------------------------------------

PennyLane offers a number of transforms that take quantum functions and alter them to represent
quantum functions of optimized circuits:

:html:`<div class="summary-table">`

.. autosummary::
    :nosignatures:

    ~pennylane.transforms.cancel_inverses
    ~pennylane.transforms.commute_controlled
    ~pennylane.transforms.merge_amplitude_embedding
    ~pennylane.transforms.cancel_inverses
    ~pennylane.transforms.merge_rotations
    ~pennylane.transforms.pattern_matching
    ~pennylane.transforms.remove_barrier
    ~pennylane.transforms.single_qubit_fusion

>     ~pennylane.transforms.undo_swaps

Reviewer: ooh this looks really nice now!

    pipeline = [commute_controlled, cancel_inverses]

    @qml.qnode(dev)
    @qml.compile(pipeline=pipeline)
    def qfunc(x, y, z):
        qml.Hadamard(wires=0)
        qml.Hadamard(wires=1)
        qml.Hadamard(wires=2)
        qml.RZ(z, wires=2)
        qml.CNOT(wires=[2, 1])
        qml.RX(z, wires=0)
        qml.CNOT(wires=[1, 0])
        qml.RX(x, wires=0)
        qml.CNOT(wires=[1, 0])
        qml.RZ(-z, wires=2)
        qml.RX(y, wires=2)
        qml.PauliY(wires=2)
        qml.CZ(wires=[1, 2])
        return qml.expval(qml.PauliZ(wires=0))

>>> print(qml.draw(qfunc)(0.2, 0.3, 0.4))
0: ──H───RX(0.4)──RX(0.2)────────────────────────────┤ ⟨Z⟩
1: ──H──╭X───────────────────────────────────────╭C──┤
2: ──H──╰C────────RZ(0.4)──RZ(-0.4)──RX(0.3)──Y──╰Z──┤

.. note::

    The :class:`~.pennylane.Barrier` operator can be used to prevent blocks of code from being merged during
    compilation.

For more details on :func:`~.pennylane.compile` and the available compilation transforms, visit
`the compilation documentation
<../code/qml_transforms.html#transforms-for-circuit-compilation>`_.

Groups of commuting Pauli words
-------------------------------

> Mutually commuting Pauli words can be measured simultaneously on a quantum computer.

Author: If that changes soon, I wonder if we should leave the ambiguity here, and people can refer to the docstring for details? I fear we forget to update...

> Mutually commuting Pauli words can be measured simultaneously on a quantum computer.

Reviewer 2: And I think it's already been merged in... So leave it be.

> Mutually commuting Pauli words can be measured simultaneously on a quantum computer.

Reviewer: @mariaschuld I might suggest having:

```rest
.. autosummary::

  ~split_non_commuting
  ~hamiltonian_expand
  ~group_observables
```

here? 

Split non-commuting is this same functionality, but at the QNode level:

```python
@qml.transforms.split_non_commuting
@qml.qnode(dev)
def circuit(x):
    qml.RX(x,wires=0)
    return [qml.expval(qml.PauliX(0)), qml.expval(qml.PauliZ(0))]

>>> print(qml.draw(circuit)(0.5))
0: ──RX(0.50)─┤  <X>
0: ──RX(0.50)─┤  <Z>
```

Hamiltonian expand is the same but for Hamiltonians. It actually _should_ support QNodes! This is just an oversight, that can be fixed by adding `@qml.batch_transform` above it's definition lol:

```python
# we should do this in the library!
hamiltonian_expand = qml.batch_transform(qml.transforms.hamiltonian_expand)

@hamiltonian_expand
@qml.qnode(dev)
def circuit(x):
    qml.RX(x,wires=0)
    return qml.expval(qml.PauliX(0) + qml.PauliZ(0))

>>> print(qml.draw(circuit)(0.5))
0: ──RX(0.50)─┤  <X>
0: ──RX(0.50)─┤  <Z>
```

I think only then would I introduce `group_observables`, for those working at a low level. Because it isn't clear to me from reading this paragraph _how to use_ `group_observables` with an existing Hamiltonian or QNode.

In fact, it might be better to first describe `H.compute_grouping()` _first_?

When given an observable that is a linear combination of Pauli words, it can therefore
be useful to find such groups in order to optimize the number of circuit runs.

This can be done with the :func:`~.pennylane.grouping.group_observables` function:

>>> obs = [qml.PauliY(0), qml.PauliX(0) @ qml.PauliX(1), qml.PauliZ(1)]
>>> coeffs = [1.43, 4.21, 0.97]
>>> obs_groupings, coeffs_groupings = group_observables(obs, coeffs, 'anticommuting', 'lf')
>>> obs_groupings
[[PauliZ(wires=[1]), PauliX(wires=[0]) @ PauliX(wires=[1])],
 [PauliY(wires=[0])]]
>>> coeffs_groupings
[[0.97, 4.21], [1.43]]

For further details on measurement optimization, grouping observables through
solving the minimum clique cover problem, and other Pauli operator logic, refer to the
:doc:`/code/qml_grouping` subpackage.

Custom decompositions

> ---------------------

Reviewer: This section is much easier to read now that it is shorter!

    a separate context manager :func:`~.pennylane.transforms.set_decomposition` can be used.

Circuit cutting
---------------

Circuit cutting allows you to replace a circuit with ``N`` wires by a set of circuits with less than
``N`` wires (see also `Peng et. al <https://arxiv.org/abs/1904.00102>`_). Of course this comes with a cost: The smaller circuits
require a greater number of device executions to be evaluated.

In PennyLane, circuit cutting can be
activated by positioning :class:`~.pennylane.WireCut` operators at the desired cut locations, and
by decorating the QNode with the :func:`~.pennylane.cut_circuit` transform.

The example below shows how a three-wire circuit can be run on a two-wire device:

.. code-block:: python

    dev = qml.device("default.qubit", wires=2)

    @qml.cut_circuit
    @qml.qnode(dev)
    def circuit(x):
        qml.RX(x, wires=0)
        qml.RY(0.9, wires=1)
        qml.RX(0.3, wires=2)

        qml.CZ(wires=[0, 1])
        qml.RY(-0.4, wires=0)

        qml.WireCut(wires=1)

        qml.CZ(wires=[1, 2])

        return qml.expval(qml.grouping.string_to_pauli_word("ZZZ"))

> Instead of executing the circuit directly, it will be partitioned into
> smaller fragments according to the :class:`~.pennylane.WireCut` locations,
> and each fragment executed multiple times. PennyLane automatically combines the results
> of the fragment executions to recover the expected output of the original uncut circuit.

Reviewer 2: ```suggestion
Instead of being executed directly, the circuit will be partitioned into
smaller fragments according to the :class:`~.pennylane.WireCut` locations,
and each fragment will be executed multiple times. PennyLane automatically combines the results
of the fragment executions to recover the expected output of the original uncut circuit.
```

section on :doc:`inspecting circuits </introduction/inspecting_circuits>`.

Compilation transforms for circuit optimization
-----------------------------------------------

PennyLane offers a number of transforms that take quantum functions and alter them to represent
quantum functions of optimized circuits:

* :func:`~.pennylane.transforms.optimization.commute_controlled`: pushes commuting single-qubit
  gates through controlled operations.

* :func:`~.pennylane.transforms.optimization.cancel_inverses`: removes adjacent pairs of operations
  that cancel out.

* :func:`~.pennylane.transforms.optimization.merge_rotations`: combines adjacent rotation gates of
  the same type into a single gate, including controlled rotations.

* :func:`~.pennylane.transforms.optimization.single_qubit_fusion`: acts on all sequences of
  single-qubit operations in a quantum function, and converts each
  sequence to a single ``Rot`` gate.

* :func:`~.pennylane.transforms.optimization.pattern_matching`: optimizes a circuit given a list of patterns
  of gates that decompose the identity.

* :func:`~.pennylane.transforms.optimization.undo_swaps`: removes SWAP gates by running from right

>   to left through the circuit and changing the position of the qubits accordingly.

Author: Good point.

I forgot to mention in the description: for the life of me I cannot figure out why the links don't work. I used all structures that work in other files and checked all paths, and for some reason the :func: directive in this file still breaks.

of quantum function transforms (you can even write your own!).
For example, if we wanted to only push single-qubit gates through
controlled gates and cancel adjacent inverses, we could do:

.. code-block:: python

    from pennylane.transforms import commute_controlled, cancel_inverses
    pipeline = [commute_controlled, cancel_inverses]

    @qml.qnode(dev)
    @qml.compile(pipeline=pipeline)
    def qfunc(x, y, z):
        qml.Hadamard(wires=0)
        qml.Hadamard(wires=1)
        qml.Hadamard(wires=2)
        qml.RZ(z, wires=2)
        qml.CNOT(wires=[2, 1])
        qml.RX(z, wires=0)
        qml.CNOT(wires=[1, 0])
        qml.RX(x, wires=0)
        qml.CNOT(wires=[1, 0])
        qml.RZ(-z, wires=2)
        qml.RX(y, wires=2)
        qml.PauliY(wires=2)
        qml.CZ(wires=[1, 2])
        return qml.expval(qml.PauliZ(wires=0))

>>> print(qml.draw(qfunc)(0.2, 0.3, 0.4))
0: ──H───RX(0.4)──RX(0.2)────────────────────────────┤ ⟨Z⟩
1: ──H──╭X───────────────────────────────────────╭C──┤
2: ──H──╰C────────RZ(0.4)──RZ(-0.4)──RX(0.3)──Y──╰Z──┤

.. note::

    The :class:`~.pennylane.Barrier` operator can be used to prevent blocks of code from being merged during
    compilation.

> For more details on :func:`~.pennylane.transforms.compile` and the available compilation transforms, visit

Author: Yay, that is it: I cannot use the full path. No idea why, but I guess it gets registered in the API docs with the shortest path for import!

`the compilation documentation

> <https://pennylane.readthedocs.io/en/stable/code/qml_transforms.html#transforms-for-circuit-compilation>`_.

Author: Unfortunately your second idea does not work, since I want to link a section, and the `qml_transforms.rst` is just an autogenerating stub if I see correctly...

> <https://pennylane.readthedocs.io/en/stable/code/qml_transforms.html#transforms-for-circuit-compilation>`_.

Author: PS: I copied the link to the stable webpage from somewhere else, if it is dangerous we should systematically remove this approach

Groups of commuting Pauli words
-------------------------------

Mutually commuting Pauli words can be measured simultaneously on a quantum computer.
When given an observable that is a linear combination of Pauli words, it can therefore
be useful to find such groups in order to optimize the number of circuit runs.

This can be done with the :func:`~.pennylane.group_observables` function:

>>> obs = [qml.PauliY(0), qml.PauliX(0) @ qml.PauliX(1), qml.PauliZ(1)]
>>> coeffs = [1.43, 4.21, 0.97]
>>> obs_groupings, coeffs_groupings = group_observables(obs, coeffs, 'anticommuting', 'lf')
>>> obs_groupings
[[PauliZ(wires=[1]), PauliX(wires=[0]) @ PauliX(wires=[1])],
 [PauliY(wires=[0])]]
>>> coeffs_groupings
[[0.97, 4.21], [1.43]]

For further details on measurement optimization, grouping observables through
solving the minimum clique cover problem, and other Pauli operator logic, refer to the
:doc:`/code/qml_grouping` subpackage.

.. note::

    PennyLane offers other methods to optimize measurements, such as qubit tapering
    found in the :mod:`~.pennylane.hf.tapering` module,

>     which exploits molecular symmetries of Hamiltonians.

Author: Happy to delete :) I just wanted to highlight qubit tapering as well, but it was too specific for an entire section

>     which exploits molecular symmetries of Hamiltonians.

Reviewer: Ah, I was actually referring to the entire "Groups of commuting Pauli words" section 😆 not just the tapering part

and ``Hadamard``, and of ``Hadamard`` into
``RZ`` and ``RY``.
We define the two decompositions like so, and pass them to a device:

.. code-block:: python

    def custom_cnot(wires):
        return [
            qml.Hadamard(wires=wires[1]),
            qml.CZ(wires=[wires[0], wires[1]]),
            qml.Hadamard(wires=wires[1])
        ]

    def custom_hadamard(wires):
        return [
            qml.RZ(np.pi, wires=wires),
            qml.RY(np.pi / 2, wires=wires)
        ]

    # Can pass the operation itself, or a string
    custom_decomps = {qml.CNOT : custom_cnot, "Hadamard" : custom_hadamard}

    decomp_dev = qml.device("default.qubit", wires=3, custom_decomps=custom_decomps)
    decomp_qnode = qml.QNode(circuit, decomp_dev)

Now when we draw or run a QNode on this device, the gates will be expanded
according to our specifications:

>>> print(qml.draw(decomp_qnode, expansion_strategy="device")(weights))
0: ──RX(0.4)──────────────────────╭C──RZ(3.14)──RY(1.57)──────────────────────────╭Z──RZ(3.14)──RY(1.57)──┤ ⟨Z⟩
1: ──RX(0.5)──RZ(3.14)──RY(1.57)──╰Z──RZ(3.14)──RY(1.57)──╭C──────────────────────│───────────────────────┤

> 2: ──RX(0.6)──RZ(3.14)──RY(1.57)──────────────────────────╰Z──RZ(3.14)──RY(1.57)──╰C──────────────────────┤

Author: Used only one decomposition and turned the context manager into a note without code.

If the custom decomposition is only supposed to be used in a specific code context,
a separate context manager :func:`~.pennylane.set_decomposition` can be used:

>>> with qml.transforms.set_decomposition(custom_decomps, original_dev):
...     print(qml.draw(original_qnode, expansion_strategy="device")(weights))
0: ──RX(0.4)──────────────────────╭C──RZ(3.14)──RY(1.57)──────────────────────────╭Z──RZ(3.14)──RY(1.57)──┤ ⟨Z⟩
1: ──RX(0.5)──RZ(3.14)──RY(1.57)──╰Z──RZ(3.14)──RY(1.57)──╭C──────────────────────│───────────────────────┤
2: ──RX(0.6)──RZ(3.14)──RY(1.57)──────────────────────────╰Z──RZ(3.14)──RY(1.57)──╰C──────────────────────┤

Circuit cutting
---------------

> 

Author: The problem here was that I had no idea what the tables would need to contain in most of the modules.

Added this one :)

> 

Reviewer: I think that makes sense! Include it when there are lots of functions referenced in a subsection, don't include it otherwise :+1:

by decorating the QNode with the :func:`~.pennylane.transforms.cut_circuit` transform.

The example below shows how a three-wire circuit can be run on a two-wire device:

.. code-block:: python

    dev = qml.device("default.qubit", wires=2)

    @qml.cut_circuit
    @qml.qnode(dev)
    def circuit(x):
        qml.RX(x, wires=0)
        qml.RY(0.9, wires=1)
        qml.RX(0.3, wires=2)

        qml.CZ(wires=[0, 1])
        qml.RY(-0.4, wires=0)

        qml.WireCut(wires=1)

        qml.CZ(wires=[1, 2])

        return qml.expval(qml.grouping.string_to_pauli_word("ZZZ"))

Instead of executing the circuit directly, it will be partitioned into
smaller fragments according to the :class:`~.pennylane.WireCut` locations,
and each fragment executed multiple times. PennyLane automatically combines the results
of the fragment executions to recover the expected output of the original uncut circuit.

>>> x = np.array(0.531, requires_grad=True)
>>> circuit(0.531)
0.47165198882111165

Circuit cutting support is also differentiable:

>>> qml.grad(circuit)(x)
-0.276982865449393

Simulated quantum circuits that produce samples can be cut using
the :func:`~.pennylane.transforms.cut_circuit_mc`

> transform based on the Monte Carlo method:

Author: True, deleted the example

.. code-block:: python

    dev = qml.device("default.qubit", wires=2, shots=1000)

    @qml.cut_circuit_mc
    @qml.qnode(dev)
    def circuit(x):
        qml.RX(0.89, wires=0)
        qml.RY(0.5, wires=1)
        qml.RX(1.3, wires=2)

        qml.CNOT(wires=[0, 1])
        qml.WireCut(wires=1)
        qml.CNOT(wires=[1, 2])

        qml.RX(x, wires=0)
        qml.RY(0.7, wires=1)
        qml.RX(2.3, wires=2)
        return qml.sample(wires=[0, 2])

>>> x = 0.3
>>> circuit(x)
tensor([[1, 1],
        [0, 1],
        [0, 1],
        ...,
        [0, 1],
        [0, 1],
        [0, 1]], requires_grad=True)

The samples are drawn from the same distribution that the original

> circuit gives rise to.

Reviewer: unrelated, but I wonder if it makes sense to have a section for transforms that _change_ circuit outputs/structure that aren't necessarily compilation? e.g.,

- `insert`
- `mitigate_with_zne`

I also need to find a place (probably not here) for `metric_tensor` and `adjoint_metric_tensor`

Author: Let's add later, this is maybe just a start? 

Once this is in master we could ask people to check out the new structure and add important features as the teams see fit...

Reviewer: Makes sense! The reason `insert` was on my mind is that it came up during the papers project that people didn't realize it existed.
